File: executors/cpp/runner.cpp
/**
 * C++ code executor for Scribble.
 * This is a wrapper that compiles and executes user C++ code.
 * The actual execution is done by compiling user code with this runner.
 */

#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <chrono>
#include <cstdlib>
#include <cstring>
#include <unistd.h>
#include <sys/wait.h>
#include <sys/resource.h>

// Base64 decode
std::string base64_decode(const std::string& encoded) {
    static const std::string base64_chars =
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";

    std::string decoded;
    std::vector<int> T(256, -1);
    for (int i = 0; i < 64; i++) T[base64_chars[i]] = i;

    int val = 0, valb = -8;
    for (unsigned char c : encoded) {
        if (T[c] == -1) break;
        val = (val << 6) + T[c];
        valb += 6;
        if (valb >= 0) {
            decoded.push_back(char((val >> valb) & 0xFF));
            valb -= 8;
        }
    }
    return decoded;
}

// Get memory from cgroups
long getCgroupsMemoryPeak() {
    std::ifstream file("/sys/fs/cgroup/memory.peak");
    if (file.is_open()) {
        long bytes;
        file >> bytes;
        return bytes / 1024; // KB
    }
    return -1;
}

// Get current memory usage
long getMemoryUsage() {
    long cgroupMem = getCgroupsMemoryPeak();
    if (cgroupMem >= 0) return cgroupMem;

    // Fallback to /proc/self/status VmRSS
    std::ifstream file("/proc/self/status");
    std::string line;
    while (std::getline(file, line)) {
        if (line.substr(0, 6) == "VmRSS:") {
            std::istringstream iss(line.substr(6));
            long kb;
            iss >> kb;
            return kb;
        }
    }
    return 0;
}

// Escape JSON string
std::string escapeJson(const std::string& s) {
    std::string result;
    for (char c : s) {
        switch (c) {
            case '"': result += "\\\""; break;
            case '\\': result += "\\\\"; break;
            case '\n': result += "\\n"; break;
            case '\r': result += "\\r"; break;
            case '\t': result += "\\t"; break;
            default: result += c;
        }
    }
    return result;
}

void printError(const std::string& status, const std::string& message,
                int compilationTimeMs, int executionTimeMs) {
    std::cout << "{\"status\":\"" << status << "\",\"error_message\":\""
              << escapeJson(message) << "\",\"compilation_time_ms\":" << compilationTimeMs
              << ",\"execution_time_ms\":" << executionTimeMs
              << ",\"memory_used_kb\":0,\"tests_passed\":0,\"tests_total\":0}" << std::endl;
}

int main() {
    const char* codeB64 = std::getenv("CODE");
    const char* testCasesJson = std::getenv("TEST_CASES");

    if (!codeB64 || strlen(codeB64) == 0) {
        printError("compilation_error", "No code provided", 0, 0);
        return 0;
    }

    // Decode code
    std::string code = base64_decode(codeB64);

    // Write user code to temp file
    std::string sourceFile = "/tmp/user_code.cpp";
    std::string execFile = "/tmp/user_code";

    std::ofstream outFile(sourceFile);
    if (!outFile) {
        printError("compilation_error", "Failed to write source file", 0, 0);
        return 0;
    }

    // Add includes and wrapper
    outFile << "#include <iostream>\n";
    outFile << "#include <vector>\n";
    outFile << "#include <string>\n";
    outFile << "#include <algorithm>\n";
    outFile << "#include <cmath>\n";
    outFile << "#include <map>\n";
    outFile << "#include <set>\n";
    outFile << "#include <queue>\n";
    outFile << "#include <stack>\n";
    outFile << "using namespace std;\n\n";
    outFile << code << "\n";
    outFile.close();

    // Compile
    auto compileStart = std::chrono::high_resolution_clock::now();

    std::string compileCmd = "g++ -O2 -std=c++17 -o " + execFile + " " + sourceFile + " 2>&1";
    FILE* pipe = popen(compileCmd.c_str(), "r");
    if (!pipe) {
        printError("compilation_error", "Failed to start compiler", 0, 0);
        return 0;
    }

    std::string compileOutput;
    char buffer[128];
    while (fgets(buffer, sizeof(buffer), pipe) != nullptr) {
        compileOutput += buffer;
    }
    int compileStatus = pclose(pipe);

    auto compileEnd = std::chrono::high_resolution_clock::now();
    int compilationTimeMs = std::chrono::duration_cast<std::chrono::milliseconds>(
        compileEnd - compileStart).count();

    if (compileStatus != 0) {
        printError("compilation_error", compileOutput, compilationTimeMs, 0);
        return 0;
    }

    // For now, just report successful compilation
    // Full test execution would require parsing JSON and running the executable
    // This is a simplified version

    long memoryKb = getMemoryUsage();

    std::cout << "{\"status\":\"accepted\",\"tests_passed\":0,\"tests_total\":0,"
              << "\"compilation_time_ms\":" << compilationTimeMs
              << ",\"execution_time_ms\":0,\"total_execution_time_ms\":0,"
              << "\"memory_used_kb\":" << memoryKb
              << ",\"test_results\":[]}" << std::endl;

    return 0;
}

File: executors/python/runner.py
#!/usr/bin/env python3
"""
Python code executor for Scribble.
Reads user code and test cases from environment variables,
executes the code against test cases, and returns results as JSON.
"""

import os
import sys
import json
import time
import base64
import subprocess
from typing import Any, Dict, List

def main():
    """Main executor entry point."""

    try:
        # Get input from environment variables
        code_b64 = os.getenv('CODE', '')
        test_cases_json = os.getenv('TEST_CASES', '[]')
        problem_id = os.getenv('PROBLEM_ID', 'unknown')

        if not code_b64:
            print(json.dumps({
                'status': 'compilation_error',
                'error_message': 'No code provided',
                'execution_time_ms': 0,
                'memory_used_kb': 0,
                'tests_passed': 0,
                'tests_total': 0
            }))
            return

        # Decode code
        code = base64.b64decode(code_b64).decode('utf-8')

        # Parse test cases
        test_cases = json.loads(test_cases_json)
        if not isinstance(test_cases, list):
            test_cases = [test_cases]

        # TODO: Implement actual code execution
        # This is a placeholder that will be filled by Developer 2

        print(json.dumps({
            'status': 'accepted',
            'execution_time_ms': 10,
            'memory_used_kb': 1024,
            'tests_passed': len(test_cases),
            'tests_total': len(test_cases),
            'test_results': [
                {
                    'test_case_id': i,
                    'passed': True,
                    'actual_output': None
                } for i in range(len(test_cases))
            ]
        }))

    except Exception as e:
        print(json.dumps({
            'status': 'runtime_error',
            'error_message': str(e),
            'execution_time_ms': 0,
            'memory_used_kb': 0,
            'tests_passed': 0,
            'tests_total': 0
        }))

if __name__ == '__main__':
    main()
